File: frontend/src/app/models/thread.ts
import { Post, PopulatedPostComparator, PopulatedPost } from './post';
import { Meeting } from './meeting';
import { PostsService } from '../deck-api/posts.service';
import { MembersService } from '../deck-api/members.service';
import { FormGroup, FormControl, Validators, ValidatorFn, AbstractControl } from '@angular/forms';

export class Thread {
    id: String;
    entry: String;
    meeting: String;
    comments: String[];
    kind: String;
    status: String;
}

declare type PopulatedThreadCallback = () => void;

export class PopulatedThread {
    id: String;
    entry: PopulatedPost;
    meeting: Meeting;
    comments: PopulatedPost[];
    kind: String;
    status: String;

    constructor(
        thread: Thread,
        postsService: PostsService,
        membersService: MembersService,
        callback?: PopulatedThreadCallback
    ) {
        this.id = thread.id;
        this.comments = [];
        this.kind = thread.kind;
        this.status = thread.status;

        postsService.getPost(thread.entry).subscribe((entry: Post) => {
            this.entry = new PopulatedPost(entry, membersService);
            if (this.comments.length === thread.comments.length && this.entry) {
                callback();
            }
        });

        for (const comment of thread.comments) {
            postsService.getPost(comment).subscribe((post: Post) => {
                this.comments.push(new PopulatedPost(post, membersService));
                if (this.comments.length === thread.comments.length && this.entry) {
                    callback();
                }
            });
        }

    }
}

export function PopulatedThreadComparator(t1: PopulatedThread, t2: PopulatedThread) {
    if (t1.entry.posted < t2.entry.posted) { return 1; }
    if (t1.entry.posted > t2.entry.posted) { return -1; }
    return 0;
}

export function SortPopulatedThreads(threads: PopulatedThread[]) {
    threads.sort(PopulatedThreadComparator);
    for (const thread of threads) {
        thread.comments.sort(PopulatedPostComparator);
    }
}

export function threadKindValidator(options: string[]): ValidatorFn {
    return (control: AbstractControl): { [key: string]: any } | null => {
        for (const option of options) {
            if (control.value === option) {
                return null;
            }
        }

        return { 'invalidOption': { value: control.value } };
    };
}

export class AddThreadForm {

    form: FormGroup;
    active: boolean;
    options = ['TEMPLATE', 'TO', 'FROM', 'MEETING'];

    constructor() {
        this.active = false;

        this.form = new FormGroup({
            kind: new FormControl('', [Validators.required, Validators.minLength(1), threadKindValidator(this.options)]),
            text: new FormControl('', [Validators.required, Validators.minLength(1)]),
        });
    }

    isActive() { return this.active; }
    toggle() { this.active = !this.active; }
    value() { return this.form.value; }
    valid() { return this.form.valid; }
}

export class AddCommentToThreadForm {

    form: FormGroup;
    active: boolean;
    threadID: String;

    constructor() {
        this.active = false;

        this.form = new FormGroup({
            text: new FormControl('', [Validators.required, Validators.minLength(1)])
        });
    }

    setThread(threadID: String) { this.threadID = threadID; }
    isActive() { return this.active && this.threadID; }
    toggle() { this.active = !this.active; }
    value() { return this.form.value; }
    valid() { return this.threadID !== undefined && this.form.valid; }
}

File: frontend/lib/models/thread.dart
import 'package:frontend/models/meeting.dart';
import 'package:frontend/models/post.dart';
import 'package:frontend/services/postService.dart';
import 'package:frontend/services/meetingService.dart';

class Thread {
  final String id;
  final String entryid;
  Post? _entry;
  final String? meetingId;
  Meeting? _meeting;
  final List<String> commentIds;
  List<Post?>? _comments;
  final String kind;
  final String status;
  final PostService _postService = PostService();
  final MeetingService _meetingService = MeetingService();
  final DateTime posted;

  Thread(
      {required this.id,
      required this.entryid,
      this.meetingId,
      required this.commentIds,
      required this.kind,
      required this.status,
      required this.posted}) {}

  factory Thread.fromJson(Map<String, dynamic> json) {
    List<String> commentIds = List.from(json['comments']);

    var thread = Thread(
      id: json['id'],
      entryid: json['entry'],
      meetingId: json['meeting'],
      commentIds: commentIds,
      kind: json['kind'],
      status: json['status'],
      posted: DateTime.parse(json['posted']),
    );
    return thread;
  }

  Future<Post?> get entry async {
    if (_entry != null) {
      return _entry!;
    }
    _entry = await _postService.getPost(entryid);
    return _entry;
  }

  Future<List<Post?>> get comments async {
    if (_comments != null) {
      return _comments!;
    }
    _comments = await Future.wait(
        commentIds.map((e) async => await _postService.getPost(e)).toList());
    return _comments!;
  }

  bool get hasMeeting {
    return meetingId != null;
  }

  Future<Meeting?> get meeting async {
    if (_meeting != null) {
      return _meeting!;
    }
    _meeting = await _meetingService.getMeeting(meetingId!);
    return _meeting;
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'entry': entry,
        'meeting': meetingId,
        'kind': kind,
        'status': status,
        'comments': commentIds
      };
}
